src/xrc/src/lib.rs

```diff
@@ -470,19 +488,28 @@ impl QueriedExchangeRate {
         }
     }
 
-    /// The function checks that the relative deviation among sufficiently many rates does
-    /// not exceed 100/[RATE_DEVIATION_DIVISOR] percent.
-    fn is_valid(&self) -> bool {
+    /// The function validates the rates in the [QueriedExchangeRate] struct.
+    fn validate(self) -> Result<Self, ExchangeRateError> {
+        // Verify that there are sufficiently many rates greater than zero.
+        if median(&self.rates) == 0 {
+            return Err(ExchangeRateError::Other(OtherError {
+                code: INVALID_RATE_ERROR_CODE,
+                description: INVALID_RATE_ERROR_MESSAGE.to_string(),
+            }));
+        }
+
+        // Verify that the relative deviation among sufficiently many rates does
+        // not exceed 100/[RATE_DEVIATION_DIVISOR] percent.
         let num = self.rates.len();
         let diff = num / 2;
         for end in diff..num {
             if self.rates[end] - self.rates[end - diff]
                 <= self.rates[end - diff].saturating_div(RATE_DEVIATION_DIVISOR)
             {
-                return true;
+                return Ok(self);
             }
         }
-        false
+        Err(ExchangeRateError::InconsistentRatesReceived)
```

Comment:
I'm not a big fan of have the return on line 509. I would expect the `Ok` result would be the last line. I would suggest the following:

```rust
        if (diff..num).all(|end| {
            self.rates[end] - self.rates[end - diff]
                > self.rates[end - diff].saturating_div(RATE_DEVIATION_DIVISOR)
        }) {
            return Err(ExchangeRateError::InconsistentRatesReceived);
        }

        Ok(self)
```

Reply:
I agree, it looks nicer. Thanks for providing the code snippet! That made it very easy to change the code.   :-)